clang/tools/archgen-mlir/ArchGenMLIR.cpp

```diff
@@ -0,0 +1,927 @@
+    return llvmirAstConsumer->GetASTDeserializationListener();
+  }
+  void PrintStats() override { return llvmirAstConsumer->PrintStats(); }
+  bool shouldSkipFunctionBody(clang::Decl *D) override {
+    return llvmirAstConsumer->shouldSkipFunctionBody(D);
+  }
+};
+
+struct NoopASTConsumer : public clang::ASTConsumer {
+};
+
+class ArchGenMLIRAction : public clang::PluginASTAction {
+
+  /// Original Main actions being replaced. it usually should be a
+  /// clang::CodeGenAction. but not always
+  std::unique_ptr<clang::FrontendAction> Inner;
+  bool HasLaunchedRealConsumer = false;
+
+public:
+  clang::CodeGenAction *getInnerAsCodeGenAction() {
+    /// clang is built without RTTI so dynamic_cast cannot be used to
+    /// check if Inner is a clang::CodeGenAction. But the only
+    /// clang::FrontendAction that returns true to hasIRSupport is
+    /// clang::CodeGenAction so we used this to check
+    if (!Inner->hasIRSupport())
+      return nullptr;
+    return static_cast<clang::CodeGenAction *>(Inner.get());
+  }
+
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &CI,
+                    llvm::StringRef InFile) override {
+    /// clang::CodeGenAction should be wrapped but not other kinds of main
+    /// actions
+    if (auto *CGAct = getInnerAsCodeGenAction()) {
+      auto InnerConsumer = Inner->CreateASTConsumer(CI, InFile);
+      if (InnerConsumer) {
+        HasLaunchedRealConsumer = true;
+        return std::make_unique<ArchGenMLIRConsumer>(
+            CI, std::move(InnerConsumer),
+            CGAct->getCodeGenerator()->GetModule());
+      }
+    }
+    return std::make_unique<NoopASTConsumer>();
+  }
+
+  /// Assignments will be set on our clang::FrontendAction and not on the Inner
+  /// clang::FrontendAction So before This function copies our data to the Inner
+  void copySelfToInner() {
+    Inner->setCurrentInput(getCurrentInput(), takeCurrentASTUnit());
+    Inner->setCompilerInstance(&getCompilerInstance());
+  }
+
+  bool hasIRSupport() const override { return Inner->hasIRSupport(); }
+
+  void ExecuteAction() override {
+    if (!HasLaunchedRealConsumer)
+      return;
+    copySelfToInner();
+    return Inner->ExecuteAction();
+  }
+
+  void EndSourceFileAction() override {
+    if (HasLaunchedRealConsumer)
+      return Inner->EndSourceFileAction();
+  }
+
+  void takeMainActionToReplace(std::unique_ptr<FrontendAction> Old) override {
+    Inner = std::move(Old);
+  }
+
+  clang::PluginASTAction::ActionType getActionType() override {
+    return clang::PluginASTAction::ActionType::ReplaceAndReuseAction;
+  }
+
+  bool ParseArgs(const clang::CompilerInstance &CI,
+                 const std::vector<std::string> &args) override {
+    return true;
+  }
+
+public:
+  ArchGenMLIRAction() {}
+};
+
+} // namespace
+
+static clang::FrontendPluginRegistry::Add<ArchGenMLIRAction>
+    X("archgen-mlir", "generate approximations in mlir");
```

Comment:
```suggestion
    X("archgen-mlir", "generate approximations in MLIR");
```
actually this should be even more elaborate.